forge/ee/routes/pipeline/index.js

```diff
@@ -0,0 +1,161 @@
+const { registerPermissions } = require('../../../lib/permissions')
+const { Roles } = require('../../../lib/roles.js')
+
+module.exports = async function (app) {
+    registerPermissions({
+        'pipeline:view': { description: 'View a pipeline', role: Roles.Member },
+        'pipeline:edit': { description: 'Edit a pipeline', role: Roles.Owner },
+        'application:pipelines:create': { description: 'Create a pipeline within an application', role: Roles.Owner },
+        'application:pipelines:list': { description: 'List pipelines within an application', role: Roles.Member },
+        'application:pipelines:delete': { description: 'Delete a pipeline from an application', role: Roles.Owner }
+    })
+
+    app.addHook('preHandler', async (request, reply) => {
+        if (request.params.pipelineId) {
+            const pipelineId = request.params.pipelineId
+            request.pipeline = await app.db.models.Pipeline.byId(pipelineId)
+            if (!request.pipeline) {
+                return reply.code(404).send({ code: 'not_found', error: 'Not Found' })
+            }
+        }
+        if (request.params.applicationId) {
+            const applicationId = request.params.applicationId
+            request.application = await app.db.models.Application.byId(applicationId)
+            if (!request.application) {
+                return reply.code(404).send({ code: 'not_found', error: 'Not Found' })
+            }
+            if (request.pipeline && request.pipeline.ApplicationId !== request.application.id) {
+                return reply.code(404).send({ code: 'not_found', error: 'Not Found' })
+            }
+        } else {
+            request.application = await app.db.models.Application.byId(request.pipeline.ApplicationId)
+        }
+
+        if (request.session.User) {
+            request.teamMembership = await request.session.User.getTeamMembership(request.application.Team.id)
+            if (!request.teamMembership && !request.session.User.admin) {
+                return reply.code(404).send({ code: 'not_found', error: 'Not Found' })
+            }
+        }
+    })
+
+    /**
+     * Add a new stage to an existing Pipeline
+     * @name /api/v1/pipelines/:pipelineId/stages
+     * @memberof forge.routes.api.pipeline
+     */
+    app.post('/pipelines/:pipelineId/stages', {
+        preHandler: app.needsPermission('pipeline:edit')
+    }, async (request, reply) => {
+        const team = await request.teamMembership.getTeam()
+        const name = request.body.name?.trim() // name of the stage
+        const instanceId = request.body.instance // instance id
+
+        let stage
+        try {
+            const options = {
+                name,
+                instance: instanceId
```

Comment:
Please always include ID in option name if it's expecting an ID rather than a whole object.

```suggestion
                instanceId: instanceId
```